### DISCUSSION TITLE
Diagnostic tracer budgets

### DISCUSSION THREAD

**Discussion by hdrake** [CliMA/Oceananigans.jl#3216] [General]
Are there any examples/validation experiments that show how a closed tracer budget can be diagnosed from `Oceananigans.jl`? I realize budget terms could be approximated manually by discretizing the various operators in the budget, but this does not ensure they will be consistent with how they are computed online (which will depend on timestepping scheme, advection scheme, corrections for anisotropic grids, etc.).

It seems like @tomchor has already tackled the more [challenging problem of TKE budgets](https://github.com/tomchor/Oceanostics.jl/blob/main/src/TKEBudgetTerms.jl) in [Oceanostics](https://github.com/tomchor/Oceanostics.jl/tree/main), but I have not found a simple cell-wise tracer budget example.

**Comment by glwagner** [comment]
I don't know of any examples; it be nice to build one up! Right now this requires some source code knowledge.

The way to do this is to use `KernelFunctionOperation` to evaluate terms. For example, looking at the tracer tendency kernel, the advective flux divergence is called `div_Uc`:

https://github.com/CliMA/Oceananigans.jl/blob/fa5e280115f619d01a460f012328bd7e6d253b38/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl#L132

so we can build a `KernelFunctionOperation` to evaluate this term:

```julia
using Oceananigans.Advection: div_Uc

tracer = model.tracers.T # for flux divergence of temperature
advection_scheme = model.advection.T
velocities = model.velocities

advective_flux_divergence_op = KernelFunctionOperation{Center, Center, Center}(div_Uc, advection_scheme, velocities, tracer)
```

The operation can be computed with

```julia
advective_flux_divergence = Field(advective_flux_divergence_op)
compute!(advective_flux_divergence)
```

The syntax is a little different for hydrostatic models vs nonhydrostatic models (which don't support different advection schemes for velocities and tracers yet).

To calculate the flux itself, then we need the functions `_advective_tracer_flux_x`, `_advective_tracer_flux_y`, `_advective_tracer_flux_z`, called here:

https://github.com/CliMA/Oceananigans.jl/blob/fa5e280115f619d01a460f012328bd7e6d253b38/src/Advection/tracer_advection_operators.jl#L30-L34

those would be built with, for example,

```julia
advective_flux_x_op = KernelFunctionOperation{Face, Center, Center}(_advective_tracer_flux_x, advection_scheme, velocities.u, tracer)
```

Note the location `(Face, Center, Center)` --- one has to know where these are evaluated --- and the extraction of `velocities.u`.

We could improve the user experience here. For example, there are two functions: one "outer" function `_advective_tracer_flux_x` and one "inner" function `advective_tracer_flux_x` (this supports special behavior when we are on `ImmersedBoundaryGrid` with bathymetry, etc). Probably those roles should be reversed; ie the outer function should simply be `advective_tracer_flux_x` with no leading underscore.

**Comment by hdrake** [comment]
Thanks @glwagner, this is a helpful pointer for getting started on this. I'll give it a shot.

**Comment by tomchor** [comment]
@hdrake I may be misunderstanding your question, but note that Oceanostics has a module for tracer variance terms here: https://github.com/tomchor/Oceanostics.jl/blob/main/src/TracerVarianceBudgetTerms.jl

Is this what you need?

Although at the moment we only have `TracerVarianceTendency`, `TracerVarianceDiffusiveTerm` and `TracerVarianceDissipationRate` (the latter being specifically the dissipation rate part of the `TracerVarianceDiffusiveTerm`). You can see an example of its use in [one of the Oceanostics' docs examples](https://tomchor.github.io/Oceanostics.jl/stable/generated/kelvin_helmholtz/).

They follow the same pattern described by @glwagner, so you can also look there for examples. If you do go through the trouble of implementing these other terms, I'd really appreciate if you could share it so that we can include it in Oceanostics and make it easier for future users . (And ofc feel free to add them there yourself!)

**Reply by glwagner** [reply]
I think @hdrake is hoping to diagnose terms in the tracer budget rather than the tracer variance budget?

**Reply by hdrake** [reply]
Yes, I was just hoping to diagnose the tracer budget (not tracer variance), including the divergent (transport) terms, not just the terms that remain in a domain integral. I linked to @tomchor's TKE example rather than the variance one just because it included more terms.

**Reply by tomchor** [reply]
Ah, I see! My bad.

The point about including the terms on Oceanostics remains though! I'd be happy to include those terms in a `TracerBudgetTerm` module if you  decide to share them later
